File: apps/windows/src/Porkn.Windows/SingBoxConfigGenerator.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Porkn.Windows;

internal static class SingBoxConfigGenerator
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };

    public static string Generate(Profile profile, int localProxyPort, RoutingSettings? routingSettings = null, Profile? chainEntryProfile = null)
    {
        routingSettings ??= RoutingSettings.Default;
        var routeRules = new List<Dictionary<string, object?>>
        {
            new()
            {
                ["inbound"] = "mixed-in",
                ["action"] = "sniff",
                ["timeout"] = "1s"
            }
        };
        routeRules.AddRange(routingSettings.RouteRules().Select(rule => rule.ToDictionary(kv => kv.Key, kv => (object?)kv.Value)));

        var outbounds = new List<Dictionary<string, object?>>();
        if (chainEntryProfile is not null)
        {
            outbounds.Add(BuildOutbound(chainEntryProfile, "chain-entry"));
            outbounds.Add(BuildOutbound(profile, "proxy-out", "chain-entry"));
        }
        else
        {
            outbounds.Add(BuildOutbound(profile));
        }
        outbounds.Add(new Dictionary<string, object?> { ["type"] = "direct", ["tag"] = "direct" });
        outbounds.Add(new Dictionary<string, object?> { ["type"] = "block", ["tag"] = "block" });

        var config = new Dictionary<string, object?>
        {
            ["log"] = new Dictionary<string, object?>
            {
                ["level"] = "info",
                ["timestamp"] = true
            },
            ["dns"] = new Dictionary<string, object?>
            {
                ["servers"] = new object[]
                {
                    new Dictionary<string, object?>
                    {
                        ["type"] = "https",
                        ["tag"] = "cloudflare",
                        ["server"] = "1.1.1.1",
                        ["server_port"] = 443,
                        ["path"] = "/dns-query",
                        ["tls"] = new Dictionary<string, object?> { ["server_name"] = "cloudflare-dns.com" }
                    },
                    new Dictionary<string, object?> { ["type"] = "local", ["tag"] = "local" }
                },
                ["final"] = "cloudflare"
            },
            ["inbounds"] = new object[]
            {
                new Dictionary<string, object?>
                {
                    ["type"] = "mixed",
                    ["tag"] = "mixed-in",
                    ["listen"] = "127.0.0.1",
                    ["listen_port"] = localProxyPort
                }
            },
            ["outbounds"] = outbounds,
            ["route"] = new Dictionary<string, object?>
            {
                ["auto_detect_interface"] = true,
                ["default_domain_resolver"] = new Dictionary<string, object?>
                {
                    ["server"] = "local",
                    ["strategy"] = "prefer_ipv4"
                },
                ["rules"] = routeRules,
                ["final"] = "proxy-out"
            }
        };

        return JsonSerializer.Serialize(config, JsonOptions);
    }

    private static Dictionary<string, object?> BuildOutbound(Profile profile, string tag = "proxy-out", string? detour = null)
    {
        return profile.Protocol.ToLowerInvariant() switch
        {
            "vless" => BuildVless(profile, tag, detour),
            "trojan" => BuildTrojan(profile, tag, detour),
            "socks" => BuildSocks(profile, tag, detour),
            _ => throw new NotSupportedException($"sing-box generator does not support {profile.Protocol.ToUpperInvariant()} yet")
        };
    }

    private static Dictionary<string, object?> BuildSocks(Profile profile, string tag, string? detour)
    {
        var outbound = BaseOutbound(profile, "socks", tag, detour);
        outbound["version"] = profile.Query.GetValueOrDefault("version", "5");
        outbound["network"] = "tcp";
        if (!string.IsNullOrWhiteSpace(profile.Username)) outbound["username"] = profile.Username;
        if (!string.IsNullOrWhiteSpace(profile.Password)) outbound["password"] = profile.Password;
        return outbound;
    }

    private static Dictionary<string, object?> BuildTrojan(Profile profile, string tag, string? detour)
    {
        var outbound = BaseOutbound(profile, "trojan", tag, detour);
        outbound["password"] = profile.Username ?? throw new InvalidOperationException("Trojan password is missing");
        outbound["tls"] = Tls(profile, defaultEnabled: true);
        return outbound;
    }

    private static Dictionary<string, object?> BuildVless(Profile profile, string tag, string? detour)
    {
        var outbound = BaseOutbound(profile, "vless", tag, detour);
        outbound["uuid"] = profile.Username ?? throw new InvalidOperationException("VLESS UUID is missing");
        outbound["network"] = profile.Query.GetValueOrDefault("network", "tcp");
        outbound["packet_encoding"] = profile.Query.GetValueOrDefault("packetEncoding", "xudp");
        if (profile.Query.TryGetValue("flow", out var flow) && !string.IsNullOrWhiteSpace(flow)) outbound["flow"] = flow;
        var tls = Tls(profile);
        if (tls is not null) outbound["tls"] = tls;
        var transport = Transport(profile);
        if (transport is not null) outbound["transport"] = transport;
        return outbound;
    }

    private static Dictionary<string, object?> BaseOutbound(Profile profile, string type, string tag, string? detour)
    {
        var outbound = new Dictionary<string, object?>
        {
            ["type"] = type,
            ["tag"] = tag,
            ["server"] = profile.Host,
            ["server_port"] = profile.Port
        };
        if (!string.IsNullOrWhiteSpace(detour)) outbound["detour"] = detour;
        return outbound;
    }

    private static Dictionary<string, object?>? Tls(Profile profile, bool defaultEnabled = false)
    {
        var security = profile.Query.GetValueOrDefault("security", profile.Query.GetValueOrDefault("tls", ""));
        var enabled = defaultEnabled || security is "tls" or "reality";
        if (!enabled) return null;

        var tls = new Dictionary<string, object?>
        {
            ["enabled"] = true,
            ["server_name"] = profile.Query.GetValueOrDefault("sni", profile.Query.GetValueOrDefault("peer", profile.Host))
        };

        if (profile.Query.TryGetValue("fp", out var fp) && !string.IsNullOrWhiteSpace(fp))
        {
            tls["utls"] = new Dictionary<string, object?> { ["enabled"] = true, ["fingerprint"] = fp };
        }

        if (security == "reality")
        {
            var reality = new Dictionary<string, object?> { ["enabled"] = true };
            if (profile.Query.TryGetValue("pbk", out var publicKey)) reality["public_key"] = publicKey;
            if (profile.Query.TryGetValue("sid", out var shortId)) reality["short_id"] = shortId;
            tls["reality"] = reality;
        }

        if (profile.Query.GetValueOrDefault("allowInsecure") == "1" || profile.Query.GetValueOrDefault("insecure") == "1")
        {
            tls["insecure"] = true;
        }

        return tls;
    }

    private static Dictionary<string, object?>? Transport(Profile profile)
    {
        var type = profile.Query.GetValueOrDefault("type", profile.Query.GetValueOrDefault("net", ""));
        if (string.IsNullOrWhiteSpace(type) || type == "tcp") return null;

        var transport = new Dictionary<string, object?> { ["type"] = type };
        if (profile.Query.TryGetValue("path", out var path) && !string.IsNullOrWhiteSpace(path)) transport["path"] = path;
        if (profile.Query.TryGetValue("host", out var host) && !string.IsNullOrWhiteSpace(host))
        {
            transport["headers"] = new Dictionary<string, object?> { ["Host"] = host };
        }
        if (profile.Query.TryGetValue("serviceName", out var serviceName) && !string.IsNullOrWhiteSpace(serviceName))
        {
            transport["service_name"] = serviceName;
        }
        return transport;
    }
}

File: apps/macos/Sources/porkn/Services/SingBoxConfigGenerator.swift
import Foundation

struct SingBoxConfigGenerator {
  enum GeneratorError: LocalizedError {
    case missingPort(TunnelProfile)
    case unsupportedProtocol(ProxyProtocol)
    case missingUserID(TunnelProfile)

    var errorDescription: String? {
      switch self {
      case .missingPort(let profile): "Для \(profile.name) не указан порт"
      case .unsupportedProtocol(let proto):
        "Генератор sing-box пока не поддерживает \(proto.displayName)"
      case .missingUserID(let profile): "Для \(profile.name) не найден UUID/user id"
      }
    }
  }

  func generate(
    profile: TunnelProfile, mode: RoutingMode, routingSettings: RoutingSettings = .current,
    localProxyPort: Int = PorknManagedProxy.defaultPort, chainEntryProfile: TunnelProfile? = nil
  ) throws -> String {
    let object = try generateObject(
      profile: profile, mode: mode, routingSettings: routingSettings, localProxyPort: localProxyPort,
      chainEntryProfile: chainEntryProfile
    )
    let data = try JSONSerialization.data(
      withJSONObject: object, options: [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes])
    return String(data: data, encoding: .utf8) ?? "{}"
  }

  func generateObject(
    profile: TunnelProfile, mode: RoutingMode, routingSettings: RoutingSettings = .current,
    localProxyPort: Int = PorknManagedProxy.defaultPort, chainEntryProfile: TunnelProfile? = nil
  ) throws -> [String: Any] {
    let routeRules = routeRules(mode: mode, routingSettings: routingSettings)
    let outbounds = try outbounds(profile: profile, chainEntryProfile: chainEntryProfile)

    return [
      "log": [
        "level": "info",
        "timestamp": true,
      ],
      "dns": [
        "servers": [
          [
            "type": "https",
            "tag": "cloudflare",
            "server": "1.1.1.1",
            "server_port": 443,
            "path": "/dns-query",
            "tls": [
              "server_name": "cloudflare-dns.com"
            ],
          ],
          [
            "type": "local",
            "tag": "local",
          ],
        ],
        "final": "cloudflare",
      ],
      "inbounds": [inbound(mode: mode, localProxyPort: localProxyPort)],
      "outbounds": outbounds,
      "route": [
        "auto_detect_interface": true,
        "default_domain_resolver": [
          "server": "local",
          "strategy": "prefer_ipv4",
        ],
        "rules": routeRules,
        "final": "proxy-out",
      ],
    ]
  }

  private func routeRules(mode: RoutingMode, routingSettings: RoutingSettings) -> [[String: Any]] {
    var rules: [[String: Any]] = [
      [
        "inbound": inboundTag(mode: mode),
        "action": "sniff",
        "timeout": "1s",
      ]
    ]

    rules.append(contentsOf: routingSettings.routeRules)

    return rules
  }

  private func inboundTag(mode: RoutingMode) -> String {
    switch mode {
    case .localProxy: "mixed-in"
    case .systemTun: "tun-in"
    }
  }

  private func inbound(mode: RoutingMode, localProxyPort: Int) -> [String: Any] {
    switch mode {
    case .localProxy:
      return [
        "type": "mixed",
        "tag": "mixed-in",
        "listen": "127.0.0.1",
        "listen_port": localProxyPort,
      ]
    case .systemTun:
      return [
        "type": "tun",
        "tag": "tun-in",
        "interface_name": "porkn0",
        "address": ["172.19.0.1/30"],
        "auto_route": true,
        "strict_route": true,
        "stack": "system",
        "mtu": 9000,
      ]
    }
  }

  private func outbounds(profile: TunnelProfile, chainEntryProfile: TunnelProfile?) throws -> [[String: Any]] {
    var result: [[String: Any]] = []
    if let chainEntryProfile {
      result.append(try outbound(profile: chainEntryProfile, tag: "chain-entry"))
      result.append(try outbound(profile: profile, tag: "proxy-out", detour: "chain-entry"))
    } else {
      result.append(try outbound(profile: profile, tag: "proxy-out"))
    }
    result.append(["type": "direct", "tag": "direct"])
    result.append(["type": "block", "tag": "block"])
    return result
  }

  private func outbound(profile: TunnelProfile, tag: String, detour: String? = nil) throws -> [String: Any] {
    guard let port = profile.serverPort else { throw GeneratorError.missingPort(profile) }

    switch profile.proto {
    case .vless:
      guard let uuid = profile.primaryUser, !uuid.isEmpty else {
        throw GeneratorError.missingUserID(profile)
      }
      var object: [String: Any] = [
        "type": "vless",
        "tag": tag,
        "server": profile.serverHost,
        "server_port": port,
        "uuid": uuid,
        "network": profile.queryItems["network"] ?? "tcp",
      ]
      if let flow = profile.queryItems["flow"], !flow.isEmpty {
        object["flow"] = flow
      }
      object["packet_encoding"] = profile.queryItems["packetEncoding"] ?? "xudp"
      if let tls = tlsObject(profile: profile) {
        object["tls"] = tls
      }
      if let transport = transportObject(profile: profile) {
        object["transport"] = transport
      }
      if let detour { object["detour"] = detour }
      return object

    case .socks:
      var object: [String: Any] = [
        "type": "socks",
        "tag": tag,
        "server": profile.serverHost,
        "server_port": port,
        "version": profile.queryItems["version"] ?? "5",
        "network": "tcp",
      ]
      if let username = profile.primaryUser, !username.isEmpty { object["username"] = username }
      if let password = profile.secret, !password.isEmpty { object["password"] = password }
      if let detour { object["detour"] = detour }
      return object

    case .trojan:
      guard let password = profile.primaryUser, !password.isEmpty else {
        throw GeneratorError.missingUserID(profile)
      }
      var object: [String: Any] = [
        "type": "trojan",
        "tag": tag,
        "server": profile.serverHost,
        "server_port": port,
        "password": password,
      ]
      if let tls = tlsObject(profile: profile, defaultEnabled: true) {
        object["tls"] = tls
      }
      if let detour { object["detour"] = detour }
      return object

    default:
      throw GeneratorError.unsupportedProtocol(profile.proto)
    }
  }

  private func tlsObject(profile: TunnelProfile, defaultEnabled: Bool = false) -> [String: Any]? {
    let security = profile.queryItems["security"] ?? profile.queryItems["tls"]
    let enabled = defaultEnabled || security == "tls" || security == "reality"
    guard enabled else { return nil }

    var tls: [String: Any] = [
      "enabled": true,
      "server_name": profile.queryItems["sni"] ?? profile.queryItems["peer"] ?? profile.serverHost,
    ]

    if let fingerprint = profile.queryItems["fp"], !fingerprint.isEmpty {
      tls["utls"] = ["enabled": true, "fingerprint": fingerprint]
    }

    if security == "reality" {
      var reality: [String: Any] = ["enabled": true]
      if let publicKey = profile.queryItems["pbk"] { reality["public_key"] = publicKey }
      if let shortID = profile.queryItems["sid"] { reality["short_id"] = shortID }
      tls["reality"] = reality
    }

    if profile.queryItems["allowInsecure"] == "1" || profile.queryItems["insecure"] == "1" {
      tls["insecure"] = true
    }

    return tls
  }

  private func transportObject(profile: TunnelProfile) -> [String: Any]? {
    let type = profile.queryItems["type"] ?? profile.queryItems["net"]
    guard let type, !type.isEmpty, type != "tcp" else { return nil }

    var object: [String: Any] = ["type": type]
    if let path = profile.queryItems["path"], !path.isEmpty { object["path"] = path }
    if let host = profile.queryItems["host"], !host.isEmpty {
      object["headers"] = ["Host": host]
    }
    if let serviceName = profile.queryItems["serviceName"], !serviceName.isEmpty {
      object["service_name"] = serviceName
    }
    return object
  }
}
